### PR TITLE
Python distrub rnd

### PR DESCRIPTION
**Pull request by NikolajLJO**

### CODE REVIEW

File: python_pellet/learner.py [hunk 1]
```diff
@@ -15,156 +15,160 @@
...
+            c3 = self.agent.targetQnet.state_dict()
+            c4 = self.agent.targetEEnet.state_dict()
+
+            for _ in range(actor_count):
+                self.q_network_que.put(copy.deepcopy(c1))
+                self.e_network_que.put(copy.deepcopy(c2))
+                self.q_t_network_que.put(copy.deepcopy(c3))
+                self.e_t_network_que.put(copy.deepcopy(c4))
+
+            logging.info("Pushed networks")
```
Comment (NikolajLJO) [inline python_pellet/learner.py]: This should not all be marked changes are:

* Added should_use_rnd as parameter
* queue sie chek from full -> qsize() == max (lars says its better)
* Added timeout 0.1 to use very small timeout but still having one
* updated target update frequency from 10k -> 10 as I is updated once per full memory use

File: python_pellet/train.py [hunk 2]
```diff
@@ -6,78 +6,82 @@
...
+                             e_t_network_que,
+                             from_actor_partition_que,
+                             to_actor_partition_que,
+                             should_use_rnd))
+            for i in range(actor_count)]
+        for process in actor_list:
+            process.start()
 
+    for p in actor_list:
+        p.join()
```
Comment (NikolajLJO) [inline python_pellet/train.py:87]: Same as learner.py actual changes are:

* added should_use_rnd to learner and actor 
* added option in args five, if it is "rnd" rnd will be used

File: python_pellet/agent.py [hunk 3]
```diff
@@ -124,72 +154,113 @@ def eelearn(self, ee_memory, batch_size=1000):
...
+						index = i
 
 			visited_prime = copy.deepcopy(visited)
 
 			if visited[0][index] is None:
-				visited[0][index] = torch.tensor([partition_memory.calc_pellet_reward(partition_memory[index][1])], device=self.device)
+				visited[0][index] = torch.tensor([partition_memory.calc_pellet_reward(partition_memory[index][1])],
+												device=self.device)
 
 			return visited, visited_prime, min_distance
```
Comment (NikolajLJO) [inline python_pellet/agent.py]: Same a sothers, actual differences are:

* added should_use_rnd check, which changes calc in qlearn

File: python_pellet/agent.py [hunk 4]
```diff
@@ -124,72 +154,115 @@ def eelearn(self, ee_memory, batch_size=1000):
...
```
Comment (NikolajLJO) [inline python_pellet/agent.py]: method find_current_partition()
Same as others, actual differences are:

* added should_use_rnd check, which changes calc in qlearn

File: python_pellet/agent.py [hunk 5]
```diff
@@ -124,72 +156,115 @@ def eelearn(self, ee_memory, batch_size=1000):
...
-				if max_distance < min_distance:
-					min_distance = max_distance
-					index = i
+				aux_forward = torch.mean(self.mse(aux_forward, aux_target_forward), dim=1)
+				aux_back = torch.mean(self.mse(aux_back, aux_target_back), dim=1)
+
+				novelties = torch.max(torch.stack([aux_forward, aux_back], dim=1), 1)[0]
+
+				argmin_value, argmin_index = torch.min(novelties, 0)
+				current_partition = partition_memory[argmin_index.item()]
```
Comment (Larssvaenjensen) [inline python_pellet/agent.py]:
```suggestion
				index = argmin_index.item()
```

File: python_pellet/learner.py [hunk 6]
```diff
@@ -15,156 +15,160 @@
...
+                    while True:
+                        from_actor_partition_que.get_nowait()
+                except queue.Empty:
+                    pass
+                logging.info("Pushed partition: " + str(self.partition))
+
+            # while we have more than 10% replay memory, learn
+            # ToDO this should prob just do entire que its a frakensetein of old concepts
+            while self.replay_memory.memory and self.ee_memory:
+                self.agent.update(self.replay_memory, self.ee_memory, ee_done, should_use_rnd=should_use_rnd)
```
Comment (Larssvaenjensen) [inline python_pellet/learner.py:153]: kør det her indtil tom

File: python_pellet/learner.py [hunk 7]
```diff
@@ -15,156 +15,160 @@
...
+                i += 1
+                if not should_use_rnd and i % 10 == 0:
+                    self.agent.targetQnet = copy.deepcopy(self.agent.Qnet)
```
Comment (Larssvaenjensen) [inline python_pellet/learner.py]: qtarget skal opdateres selvom rnd er brugt

**Review by Larssvaenjensen — APPROVED** [review]

File: python_pellet/agent.py [hunk 8]
```diff
@@ -36,34 +34,58 @@ def __init__(self, action_space, nq=0.1, ne=0.1):
...
+		self.steps_since_reward = 0
+		self.non_reward_steps_before_full_eps = 500
 
 	def cast_to_device(self, tensors):
 		for tensor in tensors:
 			tensor = tensor.to(device=self.device)
 
 	def find_action(self, state, step, visited, steps_since_reward):
+		"""
+		this function finds and return an action
```
Comment (Landi29) [inline python_pellet/agent.py:61]:
```suggestion
		This function finds and returns an action
```

File: python_pellet/agent.py [hunk 9]
```diff
@@ -82,21 +105,30 @@ def qlearn(self, replay_memory, batch_size=None):
 		terminating = torch.cat(terminating).long().to(self.device)
 		targ_mc = torch.cat(targ_mc).to(self.device)
 
+		maximum_pellet_reward = []
+		for i in range(len(states)):
+			maximum_pellet_reward.append([0.1] * 100)
+		maximum_pellet_reward = torch.tensor(maximum_pellet_reward, device=self.device)
```
Comment (Landi29) [inline python_pellet/agent.py:111]: newline after for loop?

File: python_pellet/agent.py [hunk 10]
```diff
@@ -124,72 +156,115 @@ def eelearn(self, ee_memory, batch_size=1000):
 			merged.append(merge_states_for_comparason(states[i].unsqueeze(0), s_primes[i].unsqueeze(0)))
 		self.EEnet.backpropagate(self.EEnet(torch.cat(merged)), targ_mix)
 
+	def rndlearn(self, ee_memory, batch_size: int):
+		"""
+		rndlearn calucaltes the variabels used for backpropagating the q-network
```
Comment (Landi29) [inline python_pellet/agent.py:161]:
```suggestion
		rndlearn calculates the variabels used for backpropagating the q-network
```

File: python_pellet/agent.py [hunk 11]
```diff
@@ -124,72 +156,115 @@ def eelearn(self, ee_memory, batch_size=1000):
...
+		netinput = torch.cat(netinput)
+		pred = self.EEnet(netinput)
+		targ = self.targetEEnet(netinput)
+
+		self.EEnet.backpropagate(pred, targ)
+
 	def find_current_partition(self, state, partition_memory, visited):
-		'''
+		"""
 		This function findc the partition the agent is currently in,
```
Comment (Landi29) [inline python_pellet/agent.py:181]:
```suggestion
		This function finds the partition the agent is currently in,
```

File: python_pellet/learner.py [hunk 12]
```diff
@@ -15,156 +15,160 @@
...
+                         str(len(self.replay_memory.memory)) +
+                         "rm  and: " +
+                         str(len(self.ee_memory)) +
+                         "eem and: " +
+                         str(learner_replay_que.qsize()) +
+                         "r que and: " +
+                         str(learner_ee_que.qsize()) +
+                         "eeq")
+
+            # when rpelay memory is almost empty, wait until the que has a full memory size
```
Comment (Landi29) [inline python_pellet/learner.py:86]:
```suggestion
            # when rpelay memory is almost empty, wait until the queue has a full memory size
```

**Review by Landi29 — APPROVED** [review]
Very good, just address the comments before you merge.

**Pull request merged**